File: 230-KthSmallestElementInBST/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 *
 * Because we cannot know the which branch has how many children, so 
 * we must do traversal/DFS.
 *
 * About the follow up question, we can add an attribute to each node
 * which represent how many child this node have. The the question 
 * becomes a binary search problem and the complexity is O(log(h))
 *
 * Chang Li at UC San Diego
 *
 * Dec. 5, 2015
 *
 */
public class Solution {
    
    public int count;
    public int res;
    
    public int kthSmallest(TreeNode root, int k) {
        this.count = k;
        helper(root);
        return this.res;
    }
    
    public void helper(TreeNode root) {
        if(root == null) {
            return;
        }
        // Keep diving
        helper(root.left);
        // Ok, let's go back
        count --;
        if(count == 0) {
            this.res = root.val;
            return;
        }
        // Move to another branch
        helper(root.right);
    }
}

File: 230-KthSmallestElementInBST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        _, res = self.helper(root, k)
        return res
        
    def helper(self, root, k):
        if not root:
            return (0, None)
        left_cnt, res = self.helper(root.left, k)
        if res != None:
            return (-1, res)
        elif left_cnt == k - 1:
            return (-1, root.val)
        else:
            right_cnt, res = self.helper(root.right, k - left_cnt - 1)
            return (left_cnt + right_cnt + 1, res)
